libass/ass_utils.h

```diff
@@ -104,7 +104,9 @@ int numpad2align(int val);
 unsigned ass_utf8_get_char(char **str);
 unsigned ass_utf8_put_char(char *dest, uint32_t ch);
 void ass_utf16be_to_utf8(char *dst, size_t dst_size, uint8_t *src, size_t src_size);
-#ifdef __GNUC__
+#if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 4)
+    __attribute__ ((format (gnu_printf, 3, 4)))
```

Comment:
Feel free to ignore, but I'd prefer `gnu_printf` to be specific to where it's actually required; iiuc thi is only MinGW, which overrides the default `printf` with its own at least `C99` compliant implementation. Eg by using `#if defined(__MINGW32__) && (…)`; this macro is supposedly defined in both MinGW32 and MinGW-w64 regardless of target.

Reply:
I understand the apprehension to `gnu_`, and frankly I’m not a fan of this naming choice of GCC’s, either.

Like you, I don’t feel about this so strongly as to block progress. Still, here’s my take.

My first instinct was that I’d rather avoid guessing which platforms need this. Sadly, there isn’t a test to check `print format attribute == ms_printf format attribute`. What MinGW’s headers do to `printf` declaration(s) in `stdio.h` is irrelevant (and varies); the compiler itself aliases the `printf` format attribute to `ms_printf` as a platform default. Indeed, it would make sense for it to _stop_ doing this in cases when the headers supply MinGW’s own C99-compliant `printf`, although C also allows calling `printf` without including the header…

I’ve taken a deeper dive into GCC’s source code, and it appears `printf` is an alias for `ms_printf` specifically on MinGW targets, the same ones that define `__MINGW32__`, indeed. To recap, it’s `gnu_printf` on all other platforms.

IMO it would be better if GCC offered a `std_printf`, but alas. Currently, `gnu_printf` is the closest thing to that.

By the way, out-of-standards extensions are a bit orthogonal to this: both GCC and Clang are able to verify that format strings use only the specifiers defined in the currently active C standard (`-std=`) and warn if they spot any extensions used, and this applies to `gnu_printf` equally well. We may want to activate these warnings. Unfortunately, they’re activated differently in Clang and in GCC, and in GCC the only way is `-Wpedantic`, which enables a host of other warnings. Then again, maybe we _should_ fix those warnings as well. I’m currently getting only these two kinds (on non-Windows):

> format `%p` expects argument of type `void *`, but argument 4 has type `ASS_Track *`

> ass_render.c: pointers to arrays with different qualifiers are incompatible in ISO C
> ```
>          outline_update_min_transformed_x(&info->outline->outline[0], m, leftmost_x);
>                                                                       ^
> ```
> (and several more similar instances)

(And again, perhaps the _best_ way to solve this would be to stop using nonportable format strings in the first place, as much as it would hurt our pride. It would make our logging fool-proof. But I just don’t know a good way of doing that: we’d need to log 64-bit `size_t` _somehow,_ and even plain 64-bit integer modifiers aren’t perfectly portable, although better than `%zu`.)

Reply:
(Side note: looking at GCC sources, I think `ms_printf` format checks aren’t even fully correct. `%Iu`/`%Id` seems to expect `long long` instead of `size_t`/`ptrdiff_t`, and it seems to be allowed for `scanf`, for which MSDN doesn’t list it at all.)

Reply:
> To recap, it’s gnu_printf [= printf] on all other platforms.

While I trust your finding that this currently is the case for gcc, *it shouldn't be*. [GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html) define `printf` as *“the formats accepted by the system’s C runtime library”* and `gnu_printf` as acepting *“the formats accepted by the GNU C Library”*, i.e. ISO-C + POSIX (+ XSI) + GNU extensions. Consider eg [illumos' libc](https://illumos.org/man/3C/printf) or [bionic libc]() (used by Android), which do not GNU's `%m`.
But as I wrote before, feel free to ignore.

Reply:
(Neither do macOS or FreeBSD.)

You’re not wrong, but again, extensions—whether POSIX or GNU—are not the point of this attribute. For that, there is `-Wpedantic` or `-Wformat-non-iso` (see above). If you are worried that we will use a GNU extension by accident, let’s enable those warnings. I’m all for enabling those warnings.

What `format(printf)` is meant to do is warn us _if system `printf` does not support what we use,_ regardless of its standard-ness, like on Windows, where it does not support `%zu`. In a sense, it is actually more correct to use a well-known format attribute, even if GNU, in libass, to declare “our log messages use the same formats—specifically standard C99—on all platforms, and it’s your responsibility to ensure they’re parsed correctly” rather than “our log messages use whatever your platform’s `printf` happens to support, and you should pass them through to `printf`”.

Reply:
I think the big question here is: what do you actually want to achieve?

I don’t want to talk you out of your idea just because I don’t like it, but I want to make sure we’re on the same page as to what it means.

Objective | Our best course of action
-|-
Avoid using extensions to standard C | Enable corresponding GCC and Clang warnings
Simplify preprocessor logic | Use `gnu_printf` on all platforms
Make logging work regardless of platform and callback implementation | Stop logging (or manually stringize) `size_t`, use `printf` on all platforms, and enable warnings about nonstandard formats
Protect from incorrectly printed log messages on other platforms with slightly nonstandard `printf` | Use `printf` on non-Windows, and add CI for every such platform as soon as GCC learns to recognize its `printf` format
Protect from future compiler warnings on platforms with very weird `printf` | Use `gnu_printf` on all platforms and explicitly document our logging API as using standard C99 formats on all platforms
Use precisely standard C99 formats | Same as above, and enable nonstandard format warnings

(s/C99/C11/g whenever we switch to C11.)

So like, we totally could enable nonstandard format warnings and use `printf` on non-Windows in an attempt at extra future-proof portability. But is that actually what you want? It would protect us in case GCC, due to a bug, fails to catch a nonstandard format but does catch it in a system-format check on some platform distinguished in the future from `gnu_printf`. And it would give us a warning that we could try to fix. On the other hand, this would also add a warning if GCC should learn to recognize another platform’s `printf` that actually fails to support a standard format that we use. This may be both good and bad: today, we’re getting a warning like this _and intentionally suppressing it._ Maybe it’s good to see every such incompatibility before deciding whether to work around it or suppress it.

Reply:
I'd say our format strings should match our compile-baseline, which is ISO-C99.
The goal is to get good warnings about non-ISO-C99 formats without overly complex preprocessor and/or configure logic.
The problem's are that clang and gcc employ different, incompatible options to enable actual non-ISO warnings and clang doesn't support `gnu_printf`.
Using the `printf` check is per docs supposed to be closer to non-ISO-warnings as either the format is non-ISO or the system libc fails to meet our baseline which is also worth a warning. With MinGW we have the special case of the deficient system printf being overwritten by a compliant implementation *(I guess it's actually GNU99 compliant not just ISO99)* so it's safe to use `gnu_printf` here if we want to suppress a false-positive warning.
However actually [testing](https://github.com/libass/libass/files/6779407/523_fmtWarn.c.txt) on OpenIndiana (illumos) with gcc 10 and clang 9, both of them failed to recognise the unsupported `%m` :( .

.

Regarding your previous comment about  MinGW's `printf` overwrite:
> […] although C also allows calling printf without including the header…

this shouldn't be a concern for us since we're enabling `-Werror-implicit-function-declaration` for the library build.

Reply:
We don’t log to system `printf`. We log to libass’s logging callback, which is implemented by the consumer. Our _default_ consumer uses `printf`, yes (and it’s wrong on Windows in any case, because it fails to print non-ASCII when stdout is the console), but that’s not our real concern: we can freely change the default implementation. The real concern is the public logging callback API. And we have no clue as to how the consumer implements their callback: e. g. on Windows, it could be `msvcrt.dll` `printf` without `%zu`, it could be `ucrt.dll` or MinGW `printf` with `%zu`, or it could not be a `printf` at all. What I was saying earlier is that even on modern MinGW with `-std=c11`, it could be ~~an undeclared-before-use~~ a user-declared* `printf` without `#include <stdio.h>` and end up calling `msvcrt.dll`.

\* I was reminded that C99+ technically requires function declarations, and for vararg functions even prototypes.

In fact, maybe we should `sprintf` all log messages on our side and only pass `"%s", already_formatted_log_message` downstream. And optionally deprecate the formatted callback API entirely.

> Using the `printf` check is per docs supposed to be closer to non-ISO-warnings […] However actually [testing](https://github.com/libass/libass/files/6779407/523_fmtWarn.c.txt) on OpenIndiana (illumos) with gcc 10 and clang 9, both of them failed to recognise the unsupported `%m` :( .

`format(printf)` is _literally_ an alias for `gnu_printf`, unless a particular distribution applies a custom patch. I agree that this is a deficiency in GCC that may/should/won’t be fixed in the future. But if you want to be close to ISO C, this is simply the wrong tool for the job.

You could also simply consider this a documentation bug.

Supposing GCC fixes the attribute to really fit each platform precisely (although that’d require them to have innate knowledge of all libcs which they won’t want to do, especially because versions do change, so more like… supposing each platform starts to patch GCC to make the attribute fit its own libc exactly, or GCC adds another way for system `printf` to declare its own format support), it will help _portability,_ certainly, but that’s distinct from ISO C compliance.

To be clear—I do find portability a noble goal and I’m willing to use `format(printf)` for that. But I think you’re trying to use it for what it isn’t.

> Using the `printf` check[, if we get a warning, then] either the format is non-ISO or the system libc fails to meet our baseline which is also worth a warning.

Right. But if the system libc does fail to comply with ISO C, what do we do? Do we want to change our format strings, or—like on Windows—just to declare that it’s not our problem and the log _implementation_ has to work around it if it wants correct logging? Or do we want to be warned in any case and _then_ decide?

> The goal is to get good warnings about non-ISO-C99 formats without overly complex preprocessor and/or configure logic.

Ultimately, we _need_ configure logic to get warnings about non-ISO-C99 formats at this point in time. And once we do that, we stand to gain nothing from using `printf` over `gnu_printf` even if GCC were to specialize `printf` for every libc, but we complicate the preprocessor logic slightly and make it potentially more confusing.

It doesn’t have to be very complex configure logic: a couple of `AC_COMPILE_IFELSE` with custom `CFLAGS` to see whether the compiler supports `-Wformat-non-iso` and `-Wpedantic` (or `-pedantic`; I’ll have to check whether they’re equivalent and which one is older, because why not enable it everywhere at that point). But yes, extra logic.

Reply:
Oh, and the stupidest thing is that we never actually explicitly say what format strings the callback can expect. If we were to declare that they _must_ be passed to system `...printf`, then we could, in fact, fix portability and Windows formats simply by using macros from `inttypes.h`. (FYI: _all_ flavours of system `printf` on Windows support Microsoft’s custom modifiers for `long long` and `size_t`.)

Reply:
More text about a non-blocker :)

Agree that this could be documented/implemented better by GCC and also clang which just refers to GCC's docs even though their implementaions differ (and claims `-Wpedantic` will “Warn on language extensions.” yet it doesn't actually warn about format-specifier extensions).
*(eg if `printf` were to correspond to the current selected standard and `gnu_printf` for custom mechanisms supporting all extensions.)*

> We don’t log to system printf.

Jup, what I want to say is that we should target (and also document) the format strings supported by our compile baseline (so ISO-C99). I think that's reasonable as everywhere the compile baseline is fullfilled, there will also be a fitting printing mechanism for our messages readily available. And if we ever raise the baseline, this can be changed accordingly, which might be less clear with a custom format-specifier set.

> Right. But if the system libc does fail to comply with ISO C, what do we do? 

I was thinking of those warnings as a feedback for people trying to compile libass on restricted systems. So they know the system doesn't quite meet the expectations and can decide what to do about it. *(Assuming gcc/clang ever follow the behaviour from current docs or get patched by package maintainers to do so)*

> I do find portability a noble goal and I’m willing to use format(printf) for that. But I think you’re trying to use it for what it isn’t.
> […] we stand to gain nothing from using printf over gnu_printf even if GCC were to specialize printf for every libc

I'm (now) aware of its limitation. But warnings are best effort. Prefering `printf` over `gnu_printf` seems imho better because it better signifies intent *(standard ISO)* and *will* be favourable if compilers “were to specialize printf for every libc” as it could enable warnings for ISO features *missing* from the libc.
Of course, it still allows for a superset of ISO, but that requires more work and is probably out-of scope for this PR, see next paragraph.

> Ultimately, we need configure logic to get warnings about non-ISO-C99 formats at this point in time.
> […] It doesn’t have to be very complex configure logic: a couple of AC_COMPILE_IFELSE with custom CFLAGS to see […]

If we want that, then I think this should be in a separate patch series and can be used for more.
Also, it would imho better to define a macro for testing if comppiler flags are supported, this makes reusing for the different flags easier and `configure.ac` easier to read. autoconf-archives already has [`ax_append_compile_flags`](https://www.gnu.org/software/autoconf-archive/ax_append_compile_flags.html) and [`ax_check_compile_flag`](https://www.gnu.org/software/autoconf-archive/ax_check_compile_flag.html) for this purpose. Though as we probably don't want to require full autoconf-archive to be installed for development, we could simply copy those macros(+dependencies) into our `m4` folder or write our own.
This can also be used to add the compiler flags currently added (and duplicated) in the automake files, better supporting compilers not understanding gcc argument syntax (autoconf supports a couple of those).

> In fact, maybe we should sprintf all log messages on our side and only pass "%s", already_formatted_log_message downstream. And optionally deprecate the formatted callback API entirely.

Disagree. First `snprintf` not `sprintf`. Next this takes away the performance gains of level filtering and will still be more expensive even if it get's printed. Furthermore it is either prone to cutting off long messages or needs a resizable buffer (and two `snprintf` passes) which is currently avoided if printing to a file or terminal.
As already written further above, I think assuming the format-specifiers of ISO C99 to be available is a reasonable assumption.

Reply:
> by using macros from `inttypes.h`.

We already use `PRId64` in ass.c, do we not use it for other fixed-length types?
For `size_t` there isn't a directly matching macro defined in `inttypes.h`, but I guess `PRIuPTR` should be at least as wide as `size_t`? *(It can be larger for systems with segmented memory models)*. But, …

> Microsoft’s custom modifiers for long long and size_t.

Uhh, are those proprietary length-specifers used in Microsofts `inttypes.h`? I.e., can using `inttypes.h` backfire and break the intended future doc-promise by introcuding non-ISO specifiers?

Reply:
:-)

> I.e., can using `inttypes.h` backfire and break a potential future “ISO-only”-doc-promise by introducing non-ISO specifiers?

Yep. Unfortunately. But perhaps only with older toolchains.

Older MinGW used `%I64`. In contrast, current MinGW-w64 uses `%ll` when C99-compliant `printf` is in use (whether MinGW’s own or UCRT’s).

Current Visual C++ uses `%ll` (and UCRT). I’m afraid I must’ve deleted all my instances of Visual C++ that predate UCRT but already have `inttypes.h`, so I can’t check them. I wouldn’t be surprised if they used `%I64`, but I can’t verify this.

> feedback for people trying to compile libass on restricted systems.

Fair enough.

We’re making progress :-)

> If we want that, then I think this should be in a separate patch series

Agree. And I’d say we do want that, but it’s not currently a priority.

> Next this takes away the performance gains of level filtering

Right. I forgot about that. (You’re right about the performance considerations, too.)

I’m still not a fan of the actual (underspecified IMO) design though. But never mind.

Just FYI, I think we’ll have to `vsnprintf` on console-stdout Windows in our default log implementation despite all the drawbacks you listed. I don’t think there’s any other way to get Unicode to work. And indeed, this is [how mpv does it](https://github.com/mpv-player/mpv/blob/master/osdep/io.c#L300-L357), too.

Reply:
Turns out good old `mingw32`, perhaps more properly called MinGW.org even though that domain is dead and they now [live on OSDN](https://mingw.osdn.io/), has apparently come back from the dead! They’re alive enough to publish new builds and grumble about the name being a registered trademark, at least. Notably, TDM-GCC redistributes their builds in its 32-bit-only variant, which is how I found them.

And they still unconditionally `#define PRId64 "I64d"`. Perhaps more importantly, they seem to require more explicit opt-in for C99-compliant `printf`. libass itself fits because we define `_GNU_SOURCE`, but our consumer might not. But at least there’s still _a way_ to get C99-compliant format strings to work if the consumer wants to put in the effort. (Curiously, they also do some kind of shenanigans with custom `format(mingw_printf)` etc. attributes.)

Reply:
Forgot this:

> For `size_t` there isn't a directly matching macro defined in `inttypes.h`, but I guess `PRIuPTR` should be at least as wide as `size_t`? _(It can be larger for systems with segmented memory models)_.

Yep, I ~~forgot~~ neglected the fact there’s no macro for `size_t`. The most-portable way would be to use `uintmax_t`; or we could add a custom macro specifically for Windows and keep using `%zu` elsewhere and assume/pray it works everywhere.

It’s precisely because `inttypes.h` could involve a non-ISO format specifier that I said we could use it _if we declared that the log format must be passed to the platform’s own `*printf`._ We don’t, so I fully expect someone to have a custom log implementation that does no `printf` at all and supports only ISO formats.

Reply:
This is now guarded on `__MINGW32__`, and I’ve opened issue #529 about nonstandard-format-string warnings.

Reply:
And we should simply replace `PRId64` by `%lld` in `ass.c`. It’s obviously used for portability across Windows, but given that we use `%zu` elsewhere, it’s counterproductive.